api/request.go

```diff
@@ -53,20 +90,46 @@ type Normalizable interface {
 }
 
 func (r *RequestedResource) NormalizeConnection() error {
-	url, err := url.Parse(r.Vds)
-	if err != nil {
-		return core.NewInvalidArgument(err.Error())
+
+	if len(r.Vds) == 0 || r.Vds[0] == "" {
+		return core.NewInvalidArgument("No VDS url provided")
+	}
+
+	signedUrls := false
+	if len(r.Sas) == 0 || r.Sas[0] == "" {
+		// All vds urls are signed
+		r.Sas = []string{}
+		signedUrls = true
 	}
-	if strings.TrimSpace(r.Sas) == "" {
-		if url.RawQuery == "" {
-			return core.NewInvalidArgument("No valid Sas token is found in the request")
+
+	for i, urlReq := range r.Vds {
+		urlObject, err := url.Parse(urlReq)
+		if err != nil {
+			return core.NewInvalidArgument(err.Error())
+		}
+
+		if signedUrls {
+			if urlObject.RawQuery == "" {
+				return core.NewInvalidArgument(fmt.Sprintf(
+					"No valid Sas token found at the end of vds url nr %d", i+1))
+			}
+			r.Sas = append(r.Sas, urlObject.RawQuery)
+		} else {
+			if urlObject.RawQuery != "" {
+				return core.NewInvalidArgument(fmt.Sprintf(
+					"Signed urls are not accepted when providing sas-tokens. Vds url nr %d is signed", i+1))
+			}
 		}
-		r.Sas = url.RawQuery
+		urlObject.RawQuery = ""
+		urlObject.Host = urlObject.Hostname()
+		r.Vds[i] = urlObject.String()
+	}
+
+	if len(r.Vds) != len(r.Sas) {
+		return core.NewInvalidArgument(fmt.Sprintf(
+			"Number of VDS urls and sas tokens do not match. Vds: %d, Sas: %d", len(r.Vds), len(r.Sas)))
```

Comment:
This line is introduced in one commit and fixed in the next one. Can we have the fixed version in the introduction commit itself? :slightly_smiling_face: